src/json.hpp

```diff
@@ -122,6 +143,44 @@ struct has_mapped_type
         std::is_integral<decltype(detect(std::declval<T>()))>::value;
 };
 
+// taken from http://stackoverflow.com/questions/10711952/how-to-detect-existence-of-a-class-using-sfinae
+// used to determine if json_traits is defined for a given type T
+template <typename T>
+struct has_destructor
+{
+  template <typename U>
+  static std::true_type detect(decltype(std::declval<U>().~U())*);
+
+  template <typename>
+  static std::false_type detect(...);
+
+  static constexpr bool value = decltype(detect<T>(0))::value;
+};
+
+template<typename T>
+struct has_json_traits
+{
+  static constexpr bool value = has_destructor<json_traits<T>>::value;
+};
```

Comment:
Do we need all this? 

I think we can have some member function like this in `basic_json`:

```c++
// Sketch

template <typename T>
auto from_impl(T const& t, long) -> decltype(do_something_without_trait(t));
template <typename T>
auto from_impl(T const& t, int) -> decltype(json_trait<T>::from_json(t, *this));
template <typename T>
auto from(T const& t) -> decltype(from_impl(t, 0));
```

Notice how `from` calls `from_impl(t, (int)0)`, which will try to call the overload `from_impl(T, int)` (which is a better match that the overload with long). If that works, then that is the one that is called. But if in that overload `json_trait<T>::from` fails, e.g., because there is no specialization for `T`, that overload will be silently removed, and then the second overload `from_impl(T, long)` will kick in, because there is an implicit conversion from `int` to `long`. 

So what that does is "try to call the int overload, and if that fails, fall back to the long overload". You can probably use this "idiom" to easily detect whether `json_traits` implement something that you need, and fall back to some other solution if it doesn't.

(If you want to check some examples that actually work, a good example is how `begin` and `end` are implemented in range-v3, or in STL2).